pkg/cmd/browse/browse.go

```diff
@@ -0,0 +1,245 @@
+	"github.com/spf13/cobra"
+)
+
+type browser interface {
+	Browse(string) error
+}
+
+type BrowseOptions struct {
+	HttpClient func() (*http.Client, error)
+	IO         *iostreams.IOStreams
+	BaseRepo   func() (ghrepo.Interface, error)
+	Browser    browser
+
+	SelectorArg   string
+	AdditionalArg string
+
+	ProjectsFlag bool
+	WikiFlag     bool
+	SettingsFlag bool
+	BranchFlag   bool
+}
+
+type exitCode int
+
+const (
+	exitUrlSuccess    exitCode = 0
+	exitNonUrlSuccess exitCode = 1
+	exitNotInRepo     exitCode = 2
+	exitTooManyFlags  exitCode = 3
+	exitTooManyArgs   exitCode = 4
+	exitExpectedArg   exitCode = 5
+	exitInvalidCombo  exitCode = 6
+)
+
+func NewCmdBrowse(f *cmdutil.Factory) *cobra.Command {
+
+	opts := &BrowseOptions{
+		IO:         f.IOStreams,
+		HttpClient: f.HttpClient,
+		Browser:    f.Browser,
+		BaseRepo:   f.BaseRepo,
+	}
+
+	cmd := &cobra.Command{
+		Long:  "Open specific pages in the browser on GitHub",
+		Short: "Open GitHub in the browser",
+		Use:   "browse {<number> | <path>}",
+		Args:  cobra.RangeArgs(0, 2),
+		Example: heredoc.Doc(`
+			$ gh browse
+			#=> Open current repository to the default branch
+
+			$ gh browse 217
+			#=> Open issue or pull request 217
+
+			$ gh browse --settings
+			#=> Open repository settings
+
+			$ gh browse main.go:312
+			#=> Open main.go at line 312
+
+			$ gh browse main.go --branch main
+			#=> Open main.go in the main branch
+		`),
+		Annotations: map[string]string{
+			"IsCore": "true",
+			"help:arguments": heredoc.Doc(`
+				A browser location can be specified using arguments in the following format:
+				- by number for issue or pull request, e.g. "123"; or
+				- by path for opening folders and files, e.g. "cmd/gh/main.go"
+			`),
+			"help:environment": heredoc.Doc(`
+				To configure a web browser other than the default, use the BROWSER environment variable 
+			`),
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			if len(args) > 1 {
+				opts.AdditionalArg = args[1]
+			}
+
+			if len(args) > 0 {
+				opts.SelectorArg = args[0]
+			}
+			return openInBrowser(cmd, opts)
+		},
+	}
+	cmdutil.EnableRepoOverride(cmd, f)
+	cmd.Flags().BoolVarP(&opts.ProjectsFlag, "projects", "p", false, "Open repository projects")
+	cmd.Flags().BoolVarP(&opts.WikiFlag, "wiki", "w", false, "Open repository wiki")
+	cmd.Flags().BoolVarP(&opts.SettingsFlag, "settings", "s", false, "Open repository settings")
+	cmd.Flags().BoolVarP(&opts.BranchFlag, "branch", "b", false, "Select another branch by passing in the branch name")
+
+	return cmd
+}
+
+func openInBrowser(cmd *cobra.Command, opts *BrowseOptions) error {
+
+	baseRepo, err := opts.BaseRepo()
+	httpClient, _ := opts.HttpClient()
+	apiClient := api.NewClientFromHTTP(httpClient)
+	branchName, _ := api.RepoDefaultBranch(apiClient, baseRepo)
+
+	if !inRepo(err) {
+		return printExit(exitNotInRepo, cmd, opts, "")
+	}
+
+	if getFlagAmount(cmd) > 1 {
+		return printExit(exitTooManyFlags, cmd, opts, "")
```

Comment:
I would suggest to move all flag validation to the `RunE` block, where we typically process all the flags.

You can use `cmdutil.MutuallyExclusive` to simplify checking for flags that are not compatible with each other:
```go
if err := cmdutil.MutuallyExclusive("specify only one of `--foo` or `--bar`", opts.Foo, opts.Bar); err != nil {
  return err
}
```

Reply:
I think we need a little help on this one. None of our flags can be used with any other flag, which is why we went with the option: `if getFlagAmount(cmd) > 1 {`

If we do the `MutuallyExclusive` format, would we have to cross check all four flags against each other to make sure they are not going to be used together?

Regardless, we will go ahead and move the flag checking to the `RunE` block